# F1 Analysis Project
## By Ronen Kabanovsky
### 01/10/2022




## Intro

I have been an F1 fan since 1997 when Jacques Villeneuve won the Title.  As a data analyst, I started thinking about the type of data that analysts think about and work with.  What kind of questions do they answer?


## Questions
Some questions I had about F1:
- What are the biggest factors that contribute to winning?
- Does the team with the most fastest laps in a race, go on to win the race?
- Does the team with the most fastest laps in a race, go on to win a title (drivers or constructors)
- Is there a circuit or circuits where winning occurs significantly more than other circuits?


#### Metrics
- number of wins per team
- number of fastest laps per team
- number of races per team
- number of seasons per team
- average qualifying position per team
- number of pit stops per team
- number of unique winners per circuit
<p>





## Data Prep

- Data Source: Kaggle

- Data Set: [F1 race data from 1950 to 2021](https://www.kaggle.com/rohanrao/formula-1-world-championship-1950-2020)

#### Files
- circuits.csv
- constructor_results.csv
- constructor_standings.csv
- constructors.csv
- driver_standings.csv
- drivers.csv
- lap_times.csv
- pit_stops.csv
- qualifying.csv
- races.csv
- results.csv
- seasons.csv
- status.csv

#### Loading each CSV to a dataframe 
<p> I found some code to do this but it didn't work, and I couldn't figure it out,
    so I did it manually </p>

In [350]:
# import modules
import pandas as pd
 
circuits_df = pd.read_csv("circuits.csv")
constructor_results_df = pd.read_csv("constructor_results.csv")
constructor_standings_df = pd.read_csv("constructor_standings.csv")
constructors_df = pd.read_csv("constructors.csv")
driver_standings_df = pd.read_csv("driver_standings.csv")
drivers_df = pd.read_csv("drivers.csv")
lap_times_df = pd.read_csv("lap_times.csv")
pit_stops_df = pd.read_csv("pit_stops.csv")
qualifying_df = pd.read_csv("qualifying.csv")
races_df = pd.read_csv("races.csv")
results_df = pd.read_csv("results.csv")
seasons_df = pd.read_csv("seasons.csv")
status_df = pd.read_csv("status.csv")

#### Information about the data

In [351]:
circuits_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   circuitId   79 non-null     int64  
 1   circuitRef  79 non-null     object 
 2   name        79 non-null     object 
 3   location    79 non-null     object 
 4   country     79 non-null     object 
 5   lat         79 non-null     float64
 6   lng         79 non-null     float64
 7   alt         79 non-null     object 
 8   url         79 non-null     object 
dtypes: float64(2), int64(1), object(6)
memory usage: 5.7+ KB


In [352]:
circuits_df.isnull().sum().sum()  
# checking for any null values in the dataframe

0

In [353]:
type(circuits_df['circuitRef'][0])  
# we can see that circuitRef data is actually strings, despite the data type showing 'object'

str

In [354]:
type(circuits_df['name'][0])
# we can see that name data is actually strings, despite the data type showing 'object'b

str

In [355]:
circuits_df = circuits_df.convert_dtypes()  
# We use convert_dtypes() for best possible data type.  Does it for all columns and we don't have to do it manually.

In [356]:
circuits_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   circuitId   79 non-null     Int64  
 1   circuitRef  79 non-null     string 
 2   name        79 non-null     string 
 3   location    79 non-null     string 
 4   country     79 non-null     string 
 5   lat         79 non-null     Float64
 6   lng         79 non-null     Float64
 7   alt         79 non-null     string 
 8   url         79 non-null     string 
dtypes: Float64(2), Int64(1), string(6)
memory usage: 5.9 KB


In [357]:
constructor_results_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11950 entries, 0 to 11949
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   constructorResultsId  11950 non-null  int64  
 1   raceId                11950 non-null  int64  
 2   constructorId         11950 non-null  int64  
 3   points                11950 non-null  float64
 4   status                11950 non-null  object 
dtypes: float64(1), int64(3), object(1)
memory usage: 466.9+ KB


In [358]:
constructor_results_df.isnull().sum().sum() # checking for any null values in the dataframe

0

In [359]:
constructor_results_df = constructor_results_df.convert_dtypes()  

In [360]:
constructor_results_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11950 entries, 0 to 11949
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   constructorResultsId  11950 non-null  Int64  
 1   raceId                11950 non-null  Int64  
 2   constructorId         11950 non-null  Int64  
 3   points                11950 non-null  Float64
 4   status                11950 non-null  string 
dtypes: Float64(1), Int64(3), string(1)
memory usage: 513.6 KB


In [361]:
constructor_standings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12716 entries, 0 to 12715
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   constructorStandingsId  12716 non-null  int64  
 1   raceId                  12716 non-null  int64  
 2   constructorId           12716 non-null  int64  
 3   points                  12716 non-null  float64
 4   position                12716 non-null  int64  
 5   positionText            12716 non-null  object 
 6   wins                    12716 non-null  int64  
dtypes: float64(1), int64(5), object(1)
memory usage: 695.5+ KB


In [362]:
constructor_standings_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [363]:
constructor_standings_df = constructor_standings_df.convert_dtypes()  

In [364]:
constructor_standings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12716 entries, 0 to 12715
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   constructorStandingsId  12716 non-null  Int64  
 1   raceId                  12716 non-null  Int64  
 2   constructorId           12716 non-null  Int64  
 3   points                  12716 non-null  Float64
 4   position                12716 non-null  Int64  
 5   positionText            12716 non-null  string 
 6   wins                    12716 non-null  Int64  
dtypes: Float64(1), Int64(5), string(1)
memory usage: 770.0 KB


In [365]:
driver_standings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33394 entries, 0 to 33393
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   driverStandingsId  33394 non-null  int64  
 1   raceId             33394 non-null  int64  
 2   driverId           33394 non-null  int64  
 3   points             33394 non-null  float64
 4   position           33394 non-null  int64  
 5   positionText       33394 non-null  object 
 6   wins               33394 non-null  int64  
dtypes: float64(1), int64(5), object(1)
memory usage: 1.8+ MB


In [366]:
driver_standings_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [367]:
driver_standings_df = driver_standings_df.convert_dtypes()

In [368]:
driver_standings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33394 entries, 0 to 33393
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   driverStandingsId  33394 non-null  Int64  
 1   raceId             33394 non-null  Int64  
 2   driverId           33394 non-null  Int64  
 3   points             33394 non-null  Float64
 4   position           33394 non-null  Int64  
 5   positionText       33394 non-null  string 
 6   wins               33394 non-null  Int64  
dtypes: Float64(1), Int64(5), string(1)
memory usage: 2.0 MB


In [369]:
drivers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 854 entries, 0 to 853
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   driverId     854 non-null    int64 
 1   driverRef    854 non-null    object
 2   number       854 non-null    object
 3   code         854 non-null    object
 4   forename     854 non-null    object
 5   surname      854 non-null    object
 6   dob          854 non-null    object
 7   nationality  854 non-null    object
 8   url          854 non-null    object
dtypes: int64(1), object(8)
memory usage: 60.2+ KB


In [370]:
drivers_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [371]:
drivers_df = drivers_df.convert_dtypes()

In [372]:
drivers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 854 entries, 0 to 853
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   driverId     854 non-null    Int64 
 1   driverRef    854 non-null    string
 2   number       854 non-null    string
 3   code         854 non-null    string
 4   forename     854 non-null    string
 5   surname      854 non-null    string
 6   dob          854 non-null    string
 7   nationality  854 non-null    string
 8   url          854 non-null    string
dtypes: Int64(1), string(8)
memory usage: 61.0 KB


In [373]:
lap_times_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 514592 entries, 0 to 514591
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   raceId        514592 non-null  int64 
 1   driverId      514592 non-null  int64 
 2   lap           514592 non-null  int64 
 3   position      514592 non-null  int64 
 4   time          514592 non-null  object
 5   milliseconds  514592 non-null  int64 
dtypes: int64(5), object(1)
memory usage: 23.6+ MB


In [374]:
lap_times_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [375]:
lap_times_df = lap_times_df.convert_dtypes()

In [376]:
lap_times_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 514592 entries, 0 to 514591
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   raceId        514592 non-null  Int64 
 1   driverId      514592 non-null  Int64 
 2   lap           514592 non-null  Int64 
 3   position      514592 non-null  Int64 
 4   time          514592 non-null  string
 5   milliseconds  514592 non-null  Int64 
dtypes: Int64(5), string(1)
memory usage: 26.0 MB


In [377]:
pit_stops_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8828 entries, 0 to 8827
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   raceId        8828 non-null   int64 
 1   driverId      8828 non-null   int64 
 2   stop          8828 non-null   int64 
 3   lap           8828 non-null   int64 
 4   time          8828 non-null   object
 5   duration      8828 non-null   object
 6   milliseconds  8828 non-null   int64 
dtypes: int64(5), object(2)
memory usage: 482.9+ KB


In [378]:
pit_stops_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [379]:
lap_times_df = lap_times_df.convert_dtypes()

In [380]:
lap_times_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 514592 entries, 0 to 514591
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   raceId        514592 non-null  Int64 
 1   driverId      514592 non-null  Int64 
 2   lap           514592 non-null  Int64 
 3   position      514592 non-null  Int64 
 4   time          514592 non-null  string
 5   milliseconds  514592 non-null  Int64 
dtypes: Int64(5), string(1)
memory usage: 26.0 MB


In [381]:
qualifying_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9135 entries, 0 to 9134
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   qualifyId      9135 non-null   int64 
 1   raceId         9135 non-null   int64 
 2   driverId       9135 non-null   int64 
 3   constructorId  9135 non-null   int64 
 4   number         9135 non-null   int64 
 5   position       9135 non-null   int64 
 6   q1             9127 non-null   object
 7   q2             9001 non-null   object
 8   q3             8880 non-null   object
dtypes: int64(6), object(3)
memory usage: 642.4+ KB


In [382]:
qualifying_df.isnull().sum().sum()   # checking for any null values in the dataframe

397

In [383]:
qualifying_df = qualifying_df.convert_dtypes()

In [384]:
qualifying_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9135 entries, 0 to 9134
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   qualifyId      9135 non-null   Int64 
 1   raceId         9135 non-null   Int64 
 2   driverId       9135 non-null   Int64 
 3   constructorId  9135 non-null   Int64 
 4   number         9135 non-null   Int64 
 5   position       9135 non-null   Int64 
 6   q1             9127 non-null   string
 7   q2             9001 non-null   string
 8   q3             8880 non-null   string
dtypes: Int64(6), string(3)
memory usage: 696.0 KB


In [385]:
races_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1080 entries, 0 to 1079
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   raceId     1080 non-null   int64 
 1   year       1080 non-null   int64 
 2   round      1080 non-null   int64 
 3   circuitId  1080 non-null   int64 
 4   name       1080 non-null   object
 5   date       1080 non-null   object
 6   time       1080 non-null   object
 7   url        1080 non-null   object
dtypes: int64(4), object(4)
memory usage: 67.6+ KB


In [386]:
races_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [387]:
races_df = races_df.convert_dtypes()

In [388]:
races_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1080 entries, 0 to 1079
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   raceId     1080 non-null   Int64 
 1   year       1080 non-null   Int64 
 2   round      1080 non-null   Int64 
 3   circuitId  1080 non-null   Int64 
 4   name       1080 non-null   string
 5   date       1080 non-null   string
 6   time       1080 non-null   string
 7   url        1080 non-null   string
dtypes: Int64(4), string(4)
memory usage: 71.8 KB


In [389]:
results_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25399 entries, 0 to 25398
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         25399 non-null  int64  
 1   raceId           25399 non-null  int64  
 2   driverId         25399 non-null  int64  
 3   constructorId    25399 non-null  int64  
 4   number           25399 non-null  object 
 5   grid             25399 non-null  int64  
 6   position         25399 non-null  object 
 7   positionText     25399 non-null  object 
 8   positionOrder    25399 non-null  int64  
 9   points           25399 non-null  float64
 10  laps             25399 non-null  int64  
 11  time             25399 non-null  object 
 12  milliseconds     25399 non-null  object 
 13  fastestLap       25399 non-null  object 
 14  rank             25399 non-null  object 
 15  fastestLapTime   25399 non-null  object 
 16  fastestLapSpeed  25399 non-null  object 
 17  statusId    

In [390]:
results_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [391]:
results_df = results_df.convert_dtypes()

In [392]:
results_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25399 entries, 0 to 25398
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         25399 non-null  Int64  
 1   raceId           25399 non-null  Int64  
 2   driverId         25399 non-null  Int64  
 3   constructorId    25399 non-null  Int64  
 4   number           25399 non-null  string 
 5   grid             25399 non-null  Int64  
 6   position         25399 non-null  string 
 7   positionText     25399 non-null  string 
 8   positionOrder    25399 non-null  Int64  
 9   points           25399 non-null  Float64
 10  laps             25399 non-null  Int64  
 11  time             25399 non-null  string 
 12  milliseconds     25399 non-null  string 
 13  fastestLap       25399 non-null  string 
 14  rank             25399 non-null  string 
 15  fastestLapTime   25399 non-null  string 
 16  fastestLapSpeed  25399 non-null  string 
 17  statusId    

In [393]:
seasons_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73 entries, 0 to 72
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   year    73 non-null     int64 
 1   url     73 non-null     object
dtypes: int64(1), object(1)
memory usage: 1.3+ KB


In [394]:
seasons_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [395]:
seasons_df = seasons_df.convert_dtypes()

In [396]:
seasons_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73 entries, 0 to 72
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   year    73 non-null     Int64 
 1   url     73 non-null     string
dtypes: Int64(1), string(1)
memory usage: 1.3 KB


In [397]:
status_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137 entries, 0 to 136
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   statusId  137 non-null    int64 
 1   status    137 non-null    object
dtypes: int64(1), object(1)
memory usage: 2.3+ KB


In [398]:
status_df.isnull().sum().sum()   # checking for any null values in the dataframe

0

In [399]:
status_df = status_df.convert_dtypes()

In [400]:
status_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137 entries, 0 to 136
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   statusId  137 non-null    Int64 
 1   status    137 non-null    string
dtypes: Int64(1), string(1)
memory usage: 2.4 KB


Looks like there are nulls only in the <b>qualifying</b> data.
Lets take a look

In [401]:
qualifying_df

qualifyId  raceId  driverId  constructorId  number  position        q1  \
0             1      18         1              1      22         1  1:26.572   
1             2      18         9              2       4         2  1:26.103   
2             3      18         5              1      23         3  1:25.664   
3             4      18        13              6       2         4  1:25.994   
4             5      18         2              2       3         5  1:25.960   
...         ...     ...       ...            ...     ...       ...       ...   
9130       9171    1073       849              3       6        16  1:24.338   
9131       9172    1073       847              3      63        17  1:24.423   
9132       9173    1073         8             51       7        18  1:24.779   
9133       9174    1073       854            210      47        19  1:24.906   
9134       9175    1073       853            210       9        20  1:25.685   

            q2        q3  
0     1:25.187  1:26.714  
1     1:25.315  1:26.869  
2     1:25.452  1:27.079  
3     1:25.691  1:27.178  
4     1:25.518  1:27.236  
...        ...       ...  
9130        \N        \N  
9131        \N        \N  
9132        \N        \N  
9133        \N        \N  
9134        \N        \N  

[9135 rows x 9 columns]

In [402]:
qualifying_df["q2"].count()

9001

In [403]:
qualifying_df["q3"].count()

8880

- q1, everyone qualifies (9135 rows).   
- not everyone moves on to q2 (9001 rows). 
- not everyone moves on to q3 (8880 rows).

## Checking for duplicates

We run <b>duplicated()</b> to show if there are duplicates true/false.
Then, we wrap that in the same dataframe to only show true values.

Finaly, when running several lines of code, Jupyter will only show the output from the last command.  In this case, I need
to see each output seperately.

I found the <b>display</b> module and wrapped it around each line of code.

#### If our data is good, we should see no results in each output

In [404]:
from IPython.display import display

display(circuits_df[circuits_df.duplicated()])
display(constructor_results_df[constructor_results_df.duplicated()])
display(constructor_standings_df[constructor_standings_df.duplicated()])
display(constructors_df[constructors_df.duplicated()])
display(driver_standings_df[driver_standings_df.duplicated()])
display(drivers_df[drivers_df.duplicated()])
display(lap_times_df[lap_times_df.duplicated()])
display(pit_stops_df[pit_stops_df.duplicated()])
display(qualifying_df[qualifying_df.duplicated()])
display(races_df[races_df.duplicated()])
display(results_df[results_df.duplicated()])
display(seasons_df[seasons_df.duplicated()])
display(status_df[status_df.duplicated()])

Empty DataFrame
Columns: [circuitId, circuitRef, name, location, country, lat, lng, alt, url]
Index: []

Empty DataFrame
Columns: [constructorResultsId, raceId, constructorId, points, status]
Index: []

Empty DataFrame
Columns: [constructorStandingsId, raceId, constructorId, points, position, positionText, wins]
Index: []

Empty DataFrame
Columns: [constructorId, constructorRef, name, nationality, url]
Index: []

Empty DataFrame
Columns: [driverStandingsId, raceId, driverId, points, position, positionText, wins]
Index: []

Empty DataFrame
Columns: [driverId, driverRef, number, code, forename, surname, dob, nationality, url]
Index: []

Empty DataFrame
Columns: [raceId, driverId, lap, position, time, milliseconds]
Index: []

Empty DataFrame
Columns: [raceId, driverId, stop, lap, time, duration, milliseconds]
Index: []

Empty DataFrame
Columns: [qualifyId, raceId, driverId, constructorId, number, position, q1, q2, q3]
Index: []

Empty DataFrame
Columns: [raceId, year, round, circuitId, name, date, time, url]
Index: []

Empty DataFrame
Columns: [resultId, raceId, driverId, constructorId, number, grid, position, positionText, positionOrder, points, laps, time, milliseconds, fastestLap, rank, fastestLapTime, fastestLapSpeed, statusId]
Index: []

Empty DataFrame
Columns: [year, url]
Index: []

Empty DataFrame
Columns: [statusId, status]
Index: []

## Checking for data errors

So now I want to examine the actual values of the data.   While I am confident that the data source author did an amazing job
preparing this data, this is more for me to get into this habit.  So [Vopani](https://www.kaggle.com/rohanrao), please accept my apologies.

I have been caught with invalid data a few times before, because as the saying goes
<i>"Assumptions are the mother of all f*** ups!"</i>

### Upload CSV to Google Sheets

So what I want to do now is to have a quick looking at the data in google Sheets.  Particularly the cell values themselves.
stuff like blank cells, length of values, etc.  Again, I'm sure there's a way to do all this in Python, but this is not something I will be looking into just yet.

In [405]:
from IPython.display import Image
display(Image(url= "images/lap_times_error_check.jpg", width=400, height=400))
print("I uploaded the lap_times csv and filtered the data")

I uploaded the lap_times csv and filtered the data


In [406]:
display(Image(url= "images/lap_times_race_id_blanks.jpg", width=400, height=400))
print("So the blank values in the raceid columns are just blank rows.   I'll go ahead and remove this")

So the blank values in the raceid columns are just blank rows.   I'll go ahead and remove this


In [407]:
display(Image(url= "images/missing_hour_digits.jpg", width=500, height=500))
print("These times looks very odd to me at first, until I realized that the leading 0 for Hours has been dropped")

These times looks very odd to me at first, until I realized that the leading 0 for Hours has been dropped


I can also verify this by checking the dataframe:

In [408]:
lap_times_df.loc[lap_times_df['milliseconds'] == 3138842]

raceId  driverId  lap  position       time  milliseconds
25585     861       815    9         4  52:18.842       3138842

Looks like the original data source dropeed the leading zero as well.  I found the code below, I'll run it just to make sure.

In [409]:
#millis=input("Enter time in milliseconds ")
#millis = int(millis)
#seconds=(millis/1000)%60
#seconds = int(seconds)
#minutes=(millis/(1000*60))%60
#minutes = int(minutes)
#hours=(millis/(1000*60*60))%24

#print ("%d:%d:%d" % (hours, minutes, seconds))

#### I will repeat the same process for the remaining CSVs

### Summary up to now
- Questions
- Found a dataset from a reputable source (in our case, Kaggle)
- Looked at the way the data is structured (examined the number of csv's, shape of the data, which columns can contain nulls,  etc.)
- Checked the validity of the data (removed duplicates, ensured data types match expected)
- Ensure the data is error free (no input errors, unexplainable missing values, incorrect data format, etc.)

#### Now the data is ready to be analyzed.
#### At this point, one should decide on what tool to use for analysis:
- An SQL database?
- One spreadsheet with all csv's merged as one?
- R-Studio?


## Postgres SQL

#### I decided to go with Postgres and installed version 11
#### I created a table for each of the data sources and then copied the csv's over.

In [410]:
display(Image(url= "images/psql_11_copy_csv_to_table.jpg", width=500, height=500))

In [411]:
display(Image(url= "images/db_results.jpg"))

### Analysis

Going back to our metrics:

- number of wins per team
- number of fastest laps per team
- number of races per team
- number of seasons per team
- average qualifying position per team
- number of pit stops per team

I will be using CTEs to get a subset for each metric I want to measure, and then in the parent query, I will be performing the calculations.

#### Number of Wins Per Team

In [412]:
display(Image(url= "images/num_of_wins_per_team.jpg"))

#### Number of Fastest Laps Per Team

In [413]:
display(Image(url= "images/all_time_fastest_laps_per_team.jpg"))

#### Number of Races Per Team

In [414]:
display(Image(url="images/all_time_races_per_team.jpg"))

#### Number of Seasons Per Team

In [415]:
display(Image(url="images/seasons_per_team.jpg"))

#### Number of Pit Stops Per Team

In [416]:
display(Image(url="images/pit_stops_per_constructor.jpg"))

#### Joining all the CTEs
- We wrap the entire <b>with</b> statement with <b>CREATE TABLE()</b>
- our new table is called <b>f1_agg</b>

In [417]:
display(Image(url="images/agg_table.jpg"))

## Importing SQL data

In [418]:
# This code is adapted from the tutorial hosted below:
# http://www.postgresqltutorial.com/postgresql-python/connect/

import psycopg2

# Establish a connection to the database by creating a cursor object
# The PostgreSQL server must be accessed through the PostgreSQL APP or Terminal Shell

# conn = psycopg2.connect("dbname=postgres port=5432 user=postgres password=postgres")

# Or:
conn = psycopg2.connect(
    host="localhost", 
    port = 5432, 
    database="postgres", 
    user="postgres", 
    password="postgres",
    options="-c search_path=f1,public" # This is to specify the schema search order.   Since I am working with schema 
# called f1 and not public, I need to specify it first in the search list
)

# Create a cursor object
cur = conn.cursor()

# A sample query of all data from the "f1_agg" table in the "postgres" database
cur.execute("""SELECT * FROM f1.f1_agg""")
query_results = cur.fetchall()
print(query_results)


[('Ferrari', 'Italian', Decimal('6'), 65, 87, Decimal('192'), Decimal('2'), Decimal('5'), 978, Decimal('0.20')), ('McLaren', 'British', Decimal('8'), 53, 48, Decimal('135'), Decimal('2'), Decimal('8'), 861, Decimal('0.16')), ('Williams', 'British', Decimal('9'), 47, 7, Decimal('112'), Decimal('2'), Decimal('11'), 775, Decimal('0.14')), ('Tyrrell', 'British', Decimal('14'), 29, 0, Decimal('55'), Decimal('0'), Decimal('15'), 433, Decimal('0.13')), ('Renault', 'French', Decimal('9'), 24, 14, Decimal('37'), Decimal('1'), Decimal('9'), 403, Decimal('0.09')), ('Sauber', 'Swiss', Decimal('13'), 22, 3, Decimal('0'), Decimal('2'), Decimal('14'), 395, Decimal('0')), ('Team Lotus', 'British', Decimal('11'), 29, 0, Decimal('31'), Decimal('0'), Decimal('21'), 395, Decimal('0.08')), ('Minardi', 'Italian', Decimal('18'), 21, 0, Decimal('0'), Decimal('0'), Decimal('18'), 345, Decimal('0')), ('Ligier', 'French', Decimal('14'), 21, 0, Decimal('29'), Decimal('0'), Decimal('14'), 330, Decimal('0.09')), ('

#### In a more professional environment, the credentials would not be included in the code, but rather in separate files:
http://www.postgresqltutorial.com/postgresql-python/connect/

In [419]:
# https://towardsdatascience.com/python-and-postgresql-how-to-access-a-postgresql-database-like-a-data-scientist-b5a9c5a0ea43
# A function that takes in a PostgreSQL query and outputs a pandas database 
def create_pandas_table(sql_query, database = conn):
    table = pd.read_sql_query(sql_query, database)
    return table


# Utilize the create_pandas_table function to create a Pandas data frame
# Store the data as a variable
f1_agg = create_pandas_table("SELECT * FROM f1.f1_agg")


# Close the cursor and connection to so the server can allocate
# bandwidth to other requests
cur.close()
conn.close()


f1_agg

name nationality  avg_qualifying_position  \
0           Ferrari     Italian                      6.0   
1           McLaren     British                      8.0   
2          Williams     British                      9.0   
3           Tyrrell     British                     14.0   
4           Renault      French                      9.0   
..              ...         ...                      ...   
206  Arzani-Volpini     Italian                      0.0   
207          Turner    American                     23.0   
208         Veritas      German                     20.0   
209       Wetteroth    American                     28.0   
210            Alta     British                     15.0   

     num_of_seasons_per_team  all_time_fastest_laps  all_time_wins  \
0                         65                     87          192.0   
1                         53                     48          135.0   
2                         47                      7          112.0   
3                         29                      0           55.0   
4                         24                     14           37.0   
..                       ...                    ...            ...   
206                        0                      0            0.0   
207                        0                      0            0.0   
208                        0                      0            0.0   
209                        0                      0            0.0   
210                        0                      0            0.0   

     median_pit_stops_per_constructor  median_qualifying_per_constructor  \
0                                 2.0                                5.0   
1                                 2.0                                8.0   
2                                 2.0                               11.0   
3                                 0.0                               15.0   
4                                 1.0                                9.0   
..                                ...                                ...   
206                               0.0                                0.0   
207                               0.0                                0.0   
208                               0.0                                0.0   
209                               0.0                                0.0   
210                               0.0                                0.0   

     all_time_races  wins_to_races_ratio  
0               978                 0.20  
1               861                 0.16  
2               775                 0.14  
3               433                 0.13  
4               403                 0.09  
..              ...                  ...  
206               0                 0.00  
207               0                 0.00  
208               0                 0.00  
209               0                 0.00  
210               0                 0.00  

[211 rows x 10 columns]

In [420]:
f1_agg

name nationality  avg_qualifying_position  \
0           Ferrari     Italian                      6.0   
1           McLaren     British                      8.0   
2          Williams     British                      9.0   
3           Tyrrell     British                     14.0   
4           Renault      French                      9.0   
..              ...         ...                      ...   
206  Arzani-Volpini     Italian                      0.0   
207          Turner    American                     23.0   
208         Veritas      German                     20.0   
209       Wetteroth    American                     28.0   
210            Alta     British                     15.0   

     num_of_seasons_per_team  all_time_fastest_laps  all_time_wins  \
0                         65                     87          192.0   
1                         53                     48          135.0   
2                         47                      7          112.0   
3                         29                      0           55.0   
4                         24                     14           37.0   
..                       ...                    ...            ...   
206                        0                      0            0.0   
207                        0                      0            0.0   
208                        0                      0            0.0   
209                        0                      0            0.0   
210                        0                      0            0.0   

     median_pit_stops_per_constructor  median_qualifying_per_constructor  \
0                                 2.0                                5.0   
1                                 2.0                                8.0   
2                                 2.0                               11.0   
3                                 0.0                               15.0   
4                                 1.0                                9.0   
..                                ...                                ...   
206                               0.0                                0.0   
207                               0.0                                0.0   
208                               0.0                                0.0   
209                               0.0                                0.0   
210                               0.0                                0.0   

     all_time_races  wins_to_races_ratio  
0               978                 0.20  
1               861                 0.16  
2               775                 0.14  
3               433                 0.13  
4               403                 0.09  
..              ...                  ...  
206               0                 0.00  
207               0                 0.00  
208               0                 0.00  
209               0                 0.00  
210               0                 0.00  

[211 rows x 10 columns]

In [421]:
f1_agg.describe()

avg_qualifying_position  num_of_seasons_per_team  \
count               210.000000               211.000000   
mean                 13.876190                 4.265403   
std                   6.898714                 8.152457   
min                   0.000000                 0.000000   
25%                  10.000000                 1.000000   
50%                  13.500000                 2.000000   
75%                  18.000000                 4.000000   
max                  33.000000                65.000000   

       all_time_fastest_laps  all_time_wins  median_pit_stops_per_constructor  \
count             211.000000     211.000000                        211.000000   
mean                1.625592       5.113744                          0.184834   
std                10.386672      20.145101                          0.559726   
min                 0.000000       0.000000                          0.000000   
25%                 0.000000       0.000000                          0.000000   
50%                 0.000000       0.000000                          0.000000   
75%                 0.000000       0.000000                          0.000000   
max                87.000000     192.000000                          2.000000   

       median_qualifying_per_constructor  all_time_races  wins_to_races_ratio  
count                         211.000000      211.000000           211.000000  
mean                            3.151659       56.635071             0.031469  
std                             6.536035      129.477685             0.079254  
min                             0.000000        0.000000             0.000000  
25%                             0.000000        1.000000             0.000000  
50%                             0.000000        8.000000             0.000000  
75%                             0.000000       46.500000             0.000000  
max                            25.000000      978.000000             0.550000